package com.aguiabranca.ConectaGab.config.security;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpMethod;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;

@Configuration
@EnableWebSecurity
public class SecurityConfig {

    @Autowired
    private VerificarToken verificarToken;

    @Bean
    public SecurityFilterChain filterSecurityChain(
            HttpSecurity http
    ) throws Exception {
        return http
                .csrf(csrf -> csrf.disable())
                .sessionManagement(session ->
                        session
                                .sessionCreationPolicy(SessionCreationPolicy.STATELESS))
                .authorizeHttpRequests(authorize -> authorize
                        .requestMatchers(HttpMethod.POST, "/auth/login").permitAll()
                        .requestMatchers(HttpMethod.POST, "/auth/register").permitAll()
                        .requestMatchers(HttpMethod.GET, "/api/users").hasAnyRole("LIDER", "GESTOR", "OPERADOR")
                        .requestMatchers(HttpMethod.GET, "/api/ideas").hasAnyRole("LIDER", "GESTOR", "OPERADOR")
                        .requestMatchers(HttpMethod.POST, "/api/ideas").hasRole("OPERADOR")
                        .requestMatchers(HttpMethod.PUT, "/api/ideas").hasAnyRole("GESTOR", "OPERADOR")
                        .requestMatchers(HttpMethod.DELETE, "/api/ideas").hasAnyRole("LIDER", "GESTOR", "OPERADOR")
                        .requestMatchers(HttpMethod.POST, "/api/guidelines").hasRole("LIDER")
                        .requestMatchers(HttpMethod.PUT, "/api/guidelines").hasRole("LIDER")
                        .requestMatchers(HttpMethod.DELETE, "/api/guidelines").hasRole("LIDER")
                        .requestMatchers(HttpMethod.GET, "/api/guidelines").hasAnyRole("LIDER", "GESTOR", "OPERADOR")
                        .requestMatchers(HttpMethod.GET, "/api/projects").hasAnyRole("LIDER", "GESTOR")
                        .requestMatchers(HttpMethod.POST, "/api/projects").hasRole("GESTOR")
                        .requestMatchers(HttpMethod.PUT, "/api/projects").hasRole("GESTOR")
                        .requestMatchers(HttpMethod.DELETE, "/api/projects").hasRole("GESTOR")
                )
                .addFilterBefore(
                        verificarToken,
                        UsernamePasswordAuthenticationFilter.class
                )
                .build();
    }

    @Bean
    public AuthenticationManager authenticationManager(
        AuthenticationConfiguration authConfig) {
        return authConfig.getAuthenticationManager();
    }

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }

}
